text_editor.py

```diff
@@ -1,35 +1,77 @@
 """Text Editor"""
 
-# import keyboard
+# import curses
 
 class Model:
     """
     The model for the text editor.
     API supports up/down/left/right movement of cursor and insert/delete.
     TODO: Add highlighting.
+    TODO: Read/Write files.
     """
-    def __init__(self, cursor_position=0, prev_id=None,
-                 next_id=None, curr_id=0, iterator=1) -> None:
+    def __init__(self, cursor_position=0, prev_id=None, next_id=None,
+                 curr_id=0, iterator=1, window=(10,16)) -> None:
         self.text = []
+        self.cache = {}
         self.cursor_position = cursor_position
 
         self.prev_id = prev_id
         self.next_id = next_id
         self.curr_id = curr_id
         self.iterator = iterator
 
-        self.cache = {}
+        self.window_rows = [None]*window[0]
+        self.window_cols = [0, window[1]]
+
+    def update_window_cols(self) -> None:
+        """Update the window columns relative to the cursor."""
+        if self.cursor_position > self.window_cols[1]:
+            self.window_cols[0]+=1
+            self.window_cols[1]+=1
+        elif self.cursor_position < self.window_cols[0]:
+            self.window_cols[0]-=1
+            self.window_cols[1]-=1
+
+    def update_window_rows(self, operation=None) -> None:
+        """Update the window rows relative to the stored pointers in window_rows."""
```

Comment:
Why isn't the window just the top row and left-most column?

Reply:
I don't understand, I still need to keep track of the other bound in case I decide to go backwards.

Reply:
The screen is a fixed size, so it seems like you only need to know the top.
I imagine behavior such that the window only moves if you're two lines from the edge of the window, and you move down a line, or vice versa if you're two lines from the top and you move up. And in that case the window should shift so that e.g. the current line is in the middle.
And it seems like you should have a function that generates the text given an upper left corner to lower right corner. Not just the window, since you could generate the strings for the whole file by going from (0,0) to (max_row,max_col).